segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/delegationtoken/TokenVerifierImpl.java

```diff
@@ -9,55 +9,95 @@
-                            expectedLevel, resource, claims);
-                    return false;
-                }
-            } catch (JwtException e) {
-                log.warn("Claim verification failed for resource {} because {}", resource, e);
-                return false;
-            }
-        } else {
+    public boolean isTokenValid(String resource, String token, AuthHandler.Permissions expectedLevel) {
+        try {
+            verifyToken(resource, token, expectedLevel);
             return true;
+        } catch (TokenException e) {
+            log.warn(e.getMessage(), e);
+            return false;
         }
     }
 
-    private boolean validateEntry(Map.Entry<String, Object> entry, String resource) {
-        return (entry.getKey().endsWith("/") && resource.startsWith(entry.getKey()))
-                    ||  resource.startsWith(entry.getKey() + "/")
-                || entry.getKey().equals("*");
+    @Override
+    public void verifyToken(String resource, String token, AuthHandler.Permissions expectedLevel)
+            throws TokenExpiredException, InvalidTokenException, InvalidClaimException, TokenException {
+        if (isAuthEnabled) {
+
+            // All key value pairs inside the payload are returned, including standard fields such as sub (for subject),
+            // aud (for audience), iat, exp, as well as custom fields of the form "<resource> -> <permission>" set by
+            // Pravega.
+            Set<Map.Entry<String, Object>> claims =
+                    JsonWebToken.fetchClaims(token, tokenSigningKey);
+
+            Optional<Map.Entry<String, Object>> matchingClaim = claims.stream()
+                    .filter(entry -> entryMatchesResource(entry, resource) &&
+                                     expectedLevel.compareTo(AuthHandler.Permissions.valueOf(entry.getValue().toString())) <= 0)
+                    .findFirst();
+
+            if (!matchingClaim.isPresent()) {
+                log.trace(String.format("No matching claim found for resource [%s] and permission [%s] in token [%s].",
+                        resource, expectedLevel.toString(), token));
+
+                throw new InvalidClaimException(String.format(
+                        "No matching claim found for resource: [%s] and permission: [%s] in the delegation token.",
+                        resource, expectedLevel.toString()));
+
+            }
+        }
+    }
+
+    private boolean entryMatchesResource(Map.Entry<String, Object> entry, String resource) {
+        return resource.equals(entry.getKey()) // exact match
+
+               /*
+                * Examples of matches:
+                *   - Entry = "abc/, ...", resource = "abc/xyx"
+                */
+               || entry.getKey().endsWith("/") && resource.startsWith(entry.getKey())
+
+               /*
+                * Examples of matches:
+                *    1) Entry = "_system/_requeststream, READ_UPDATE", resource = "_system/_requeststream/0.#epoch.0"
```

Comment:
Nit: there is only one element in the list, so it does not make sense to number with `1)`.

Reply:
:) I agree. I have modified this part to collate all the comments now.